File: 576. Out of Boundary Paths/findPaths.java
class Solution {
    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {

        int dp[][][] = new int[m][n][maxMove + 1];
        int mod = 1000000000 + 7;

        for (int k = 1; k <= maxMove; k++) {
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {

                    int delrow[] = { 1, 0, -1, 0 };
                    int delcol[] = { 0, -1, 0, 1 };
                    for (int d = 0; d < 4; d++) {
                        int x = i + delrow[d];
                        int y = j + delcol[d];

                        if (x < 0 || x >= m || y < 0 || y >= n)
                            dp[i][j][k] = (dp[i][j][k] + 1) % mod;
                        else
                            dp[i][j][k] = (dp[i][j][k] + dp[x][y][k - 1]) % mod;
                    }
                }
            }
        }
        return dp[startRow][startColumn][maxMove];
    }
}

File: 576. Out of Boundary Paths/findPaths.cpp
class Solution
{
public:
    int M, N;
    int dp[51][51][51];
    int mod = 1e9 + 7;
    int solve(int row, int col, int moves)
    {
        if (row < 0 || row >= M || col < 0 || col >= N)
        {
            return 1;
        }
        if (moves <= 0)
            return 0;
        if (dp[row][col][moves] != -1)
            return dp[row][col][moves];
        long result = 0;
        int dRow[] = {1, 0, -1, 0};
        int dCol[] = {0, -1, 0, 1};
        for (int i = 0; i < 4; i++)
        {
            int nRow = row + dRow[i];
            int nCol = col + dCol[i];

            result = (result + solve(nRow, nCol, moves - 1)) % mod;
        }
        return dp[row][col][moves] = result;
    }
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn)
    {
        M = m;
        N = n;
        memset(dp, -1, sizeof(dp));
        return solve(startRow, startColumn, maxMove);
    }
};